### ISSUE TITLE
kiwi is failing to set SELinux labels correctly and failing to detect setfiles supported options

### ISSUE THREAD

**Issue by Conan-Kudo** [OSInside/kiwi#2192]
## Problem description

@davide125 noticed that our Fedora Asahi Remix images were not working with SELinux enforcing enabled, with all kinds of labeling errors on the image. Some spelunking revealed that kiwi seems to fail to detect how to use `setfiles` correctly.

From the output:
```
[ INFO    ]: 01:41:11 | Processing SELinux file security contexts
[ WARNING ]: 01:41:11 | Could not parse setfiles output
```

From the `image-root.log`:
```
INFO: 01:41:11 | Processing SELinux file security contexts
DEBUG: 01:41:11 | EXEC: [chroot /home/fedora/ngompa/test/kiwi-descriptions/outdir/build/image-root setfiles --help]
WARNING: 01:41:11 | Could not parse setfiles output
DEBUG: 01:41:11 | EXEC: [chroot /home/fedora/ngompa/test/kiwi-descriptions/outdir/build/image-root setfiles -F -p -c /etc/selinux/targeted/policy/policy.33 -e /proc -e /sys -e /dev /etc/selinux/targeted/contexts/files/file_contexts /]
```

I suspect the fact `setfiles` doesn't handle `--help` might be throwing things off? It doesn't choke on `-h`.

(as an aside, it's very annoying that these tools are so poorly documented and versioned...)

## Expected behavior

It should be able to read the `setfiles` help and detect the supported options correctly.

## Steps to reproduce the behavior

1. Boot up Fedora Linux 36 AArch64
2. Install kiwi and distribution-gpg-keys :`sudo dnf install kiwi distribution-gpg-keys`
3. Clone the descriptions down from https://pagure.io/fedora-asahi/kiwi-descriptions.git
4. Build an image (e.g. KDE Workstation): `sudo kiwi-ng --type=oem --profile=Workstation-KDE --color-output system build --description ./ --target-dir ./outdir`

## OS and Software information

* KIWI version: 9.24.44
* Operating system host version: Fedora Linux 36
* Operating system target version: Fedora Linux 38 (Rawhide)
* Open Build Service version (N/A if not using OBS): N/A
* Koji version (N/A if not using Koji): N/A

**Comment by Conan-Kudo** [comment]
Monkey-patching the code to use `setfiles -h` doesn't fix the problem either:

```
DEBUG: 02:13:34 | EXEC: [chroot /home/fedora/ngompa/test/kiwi-descriptions/outdir/build/image-root setfiles -h]
WARNING: 02:13:34 | Could not parse setfiles output
DEBUG: 02:13:34 | EXEC: [chroot /home/fedora/ngompa/test/kiwi-descriptions/outdir/build/image-root setfiles -F -p -c /etc/selinux/targeted/policy/policy.33 -e /proc -e /sys -e /dev /etc/selinux/targeted/contexts/files/file_contexts /]
```

**Comment by Conan-Kudo** [comment]
For the moment, we're working around this by [making SELinux operate in permissive mode](https://pagure.io/fedora-asahi/kiwi-descriptions/c/ad92f970f4bf4a2a825bcf774cda119118754a23), but I'd like to see this fixed so we can revert that.

**Comment by schaefi** [comment]
yeah, we have to fix this. Interesting is that I have tested the current code with I think Fedora 35 and it all worked

**Comment by schaefi** [comment]
Out of curiosity what help information does setfiles show you ? (either on -h or --help)

The reason is in kiwi we do the following:

```python
# setfiles doesn't come with a stable command API and older versions
# needs a different invocation syntax. On older versions the usage
# information explicitly lists "setfiles -c policyfile" which is not
# present in newer versions. As setfiles doesn't come with a simple
# --version option, checking for this extra element in the usage
# was the only pointer I could come up with to differentiate the
# call options.
if CommandCapabilities.has_option_in_help(
    'setfiles', 'setfiles -c policyfile', ['--help'],
    root=self.root_dir, raise_on_error=False
):

    ...
```

and I agree the warning message should at least contain the string it searches for
We are going to fix this with a PR for this one too

**Comment by davide125** [comment]
On Fedora 36:

```
$ setfiles --help
setfiles: invalid option -- '-'
usage:  setfiles [-diIDlmnpqvEFWT] [-e excludedir] [-r alt_root_path] [-c policyfile] spec_file pathname...
usage:  setfiles [-diIDlmnpqvEFWT] [-e excludedir] [-r alt_root_path] [-c policyfile] spec_file -f filename
usage:  setfiles -s [-diIDlmnpqvFWT] spec_file
$ setfiles -h
usage:  setfiles [-diIDlmnpqvEFWT] [-e excludedir] [-r alt_root_path] [-c policyfile] spec_file pathname...
usage:  setfiles [-diIDlmnpqvEFWT] [-e excludedir] [-r alt_root_path] [-c policyfile] spec_file -f filename
usage:  setfiles -s [-diIDlmnpqvFWT] spec_file
```

**Comment by schaefi** [comment]
Thanks, yes that explains why the option based condition checking failed. There is still no "setfiles --version" or something which would allow us to differentiate on the different call types right ?

**Comment by schaefi** [comment]
I'm looking into this right now. So the option checking checks for the "old" usage and if that does not apply the "new" call options are used. This means according to the log you see the following setfiles call:

```
chroot /home/fedora/ngompa/test/kiwi-descriptions/outdir/build/image-root setfiles -F -p -c /etc/selinux/targeted/policy/policy.33 -e /proc -e /sys -e /dev /etc/selinux/targeted/contexts/files/file_contexts /
```

I was wondering what's wrong with the call ?

> I suspect the fact setfiles doesn't handle --help might be throwing things off? It doesn't choke on -h

I think this is not the reason of the problem. The option checking is done for older versions of setfiles
which does not apply here. If the wrong option set would be used this would have caused setfiles to fail
but that doesn't seem to be the case ?

My understanding is that you could build the image but the setfiles result is not satisfying which leads you
to move to permissive mode as a workaround.

I'm far away from being an selinux expert. So I need a bit of help to understand the real issue.
I think it's not the caller options

Thanks

**Comment by davidcassany** [comment]
@Conan-Kudo I am wondering if this error could be fixed by simply relabelling at boot. From what I read I'd expect that adding a  `/.autorelabel` file in the root-tree should fix the problem. In that case I'd expect the system to relabel everything during the firstboot.

Like Marcus I am not a SELinux expert, no even close, but I find weird that this fails on specific aarch64 arch, we are not applying arch specific config in SELinux as far as I remember. So in case it is only failing on arm I believe there is either an issue with the selinux packages for arm or either KIWI should apply an arch specific recipe for SELinux (I really hope it will not be the case).

**Comment by Conan-Kudo** [comment]
Hmm, there may be something else funky going on then.

@zpytela do you have any idea, maybe?

**Comment by zpytela** [comment]
IIUC, there are two independent problems, distinguishing between setfiles versions and some problem with selinux policy.

The former can possibly be addressed by checking policycoreutils version, but you can RFE some --version switch. For the latter I can't see any relevant information like audit.log or ausearch command output to check.

**Comment by Conan-Kudo** [comment]
Well, we're doing an offline policy apply, so I don't think we'd have either of those log files.

**Comment by zpytela** [comment]
I don't know how could I help you then. Can you at least nail down the command which fails in SELinux enforcing and does not in permissive, or share some other details? There still can be entries in journal.

**Comment by davide125** [comment]
I can confirm this is purely a labeling issue. On a system running the kiwi-built image in permissive mode, if I `touch /.autorelabel`, reboot then switch to enforcing, everything is good and `audit2allow -b` doesn't show any output.

**Comment by davide125** [comment]
We're working this around with https://pagure.io/fedora-asahi/kiwi-descriptions/c/6edbf5bcb9a3416ac571b1cfb43da526badb8771?branch=rawhide for now but it'd be great to get it fixed properly, as autorelabel requires an extra reboot.

**Comment by schaefi** [comment]
@Conan-Kudo @davidcassany I'm lost on this one. I don't see that kiwi is calling setfiles wrong. So what should we fix here ?

**Comment by Conan-Kudo** [comment]
We're still trying to figure out what's going on. You're right the command is theoretically correct.

**Comment by schaefi** [comment]
@Conan-Kudo hope it's ok if I assign your name on this one. It came up together with the "Asahi" project and seems arch specific. Thanks

**Comment by Conan-Kudo** [comment]
I just had a brain blast as I'm about to fall asleep, so I'm writing this so it doesn't get forgotten. Are we applying policy rules before or after we sync all the files to the disk image? One difference between us and livecd-tools / livemedia-creator / etc. is that most of our operations don't occur on the blank disk, but instead on a filesystem that triggers SELinux label inheritance from the host.

What if we moved SELinux relabeling to after we sync all the files to the image target?

**Comment by Conan-Kudo** [comment]
This should be fixed with 9628abd7f4640740300936694db6013078def81e.

**Comment by Conan-Kudo** [comment]
The other SELinux issue is already tracked in #2350 .
